app/src/main/java/org/thoughtcrime/securesms/util/IP2Country.kt

```diff
@@ -66,64 +54,45 @@ class IP2Country private constructor(private val context: Context) {
                 populateCacheIfNeeded()
             }
         }
-        LocalBroadcastManager.getInstance(context).registerReceiver(pathsBuiltEventReceiver, IntentFilter("pathsBuilt"))
+        LocalBroadcastManager.getInstance(context)
+            .registerReceiver(pathsBuiltEventReceiver, IntentFilter("pathsBuilt"))
     }
 
-    // TODO: Deinit?
-    // endregion
-
-    // region Implementation
-    private fun loadFile(fileName: String): File {
-        val directory = File(context.applicationInfo.dataDir)
-        val file = File(directory, fileName)
-        if (directory.list().contains(fileName)) { return file }
-        val inputStream = context.assets.open("csv/$fileName")
-        val outputStream = FileOutputStream(file)
-        val buffer = ByteArray(1024)
-        while (true) {
-            val count = inputStream.read(buffer)
-            if (count < 0) { break }
-            outputStream.write(buffer, 0, count)
+    private fun populateCacheIfNeeded() {
+        ThreadUtils.queue {
+            // Preload if needed
+            OnionRequestAPI.paths.asSequence().flatten().map { it.ip }.forEach(::cacheCountryForIP)
+            Broadcaster(context).broadcast("onionRequestPathCountriesLoaded")
+            Log.d("Loki", "Finished preloading onion request path countries.")
         }
-        inputStream.close()
-        outputStream.close()
-        return file
     }
 
-    private fun cacheCountryForIP(ip: String): String? {
-
-        // return early if cached
-        countryNamesCache[ip]?.let { return it }
-
-        val comps = ipv4ToCountry.asSequence()
+    // TODO: Deinit?
 
-        val bestMatchCountry = comps.lastOrNull { it.key <= Ipv4Int(ip)  }?.let { (_, code) ->
-            if (code != null) {
-                countryToNames[code]
-            } else {
-                null
+    private fun loadFile(fileName: String): File =
+        File(context.applicationInfo.dataDir).let { directory ->
+            File(directory, fileName).also { file ->
+                if (directory.list()?.contains(fileName) != true) {
+                    context.assets.open("csv/$fileName").use { inputStream ->
+                        FileOutputStream(file).use(inputStream::copyTo)
+                    }
+                }
             }
         }
 
-        if (bestMatchCountry != null) {
-            countryNamesCache[ip] = bestMatchCountry
-            return bestMatchCountry
-        } else {
-            Log.d("Loki","Country name for $ip couldn't be found")
-        }
-        return null
-    }
-
-    private fun populateCacheIfNeeded() {
-        ThreadUtils.queue {
-            OnionRequestAPI.paths.iterator().forEach { path ->
-                path.iterator().forEach { snode ->
-                    cacheCountryForIP(snode.ip) // Preload if needed
-                }
+    /**
+     * get the country name for the given IP address.
+     *
+     * Cache the result even if null, as this will improve subsequent searches.
```

Comment:
Are you sure this is true?
`getOrPut`: _Returns the value for the given key if the value is present **and not null**_
Won't null value always go through the put then?